### PR TITLE
feat(component): Adds SmartScrollbar to ui-next - OHIF-2558

### PR DESCRIPTION
**Pull request by dan-rukas**
### Context

This PR adds a SmartScrollbar component to `@ohif/ui-next` that visualizes slice loading progress, viewed state, and scroll position in viewports.

The component is pure UI — it has no Cornerstone or viewer dependencies. It accepts a slice index, total count, loading state, and `Uint8Array`s of marked positions (loaded/viewed), and handles all measurement, animation, pointer interaction, and visual state internally.

Preview the component, animations, and behaviors here: https://ohif-scroll-design-final.netlify.app/

SmartScrollbar-visual-states

### Changes & Results

**New files** — 8 component files in `platform/ui-next/src/components/SmartScrollbar/`:

| File | Role |
|------|------|
| `SmartScrollbar.tsx` | Root — dual context providers (layout + scroll), ResizeObserver, pointer handling, contraction state |
| `SmartScrollbarTrack.tsx` | Dot-grid background, visible during loading |
| `SmartScrollbarFill.tsx` | Pixel-aligned colored bars for contiguous marked runs (loaded and viewed) |
| `SmartScrollbarIndicator.tsx` | Pill-shaped position marker, aligned to pixel buckets |
| `SmartScrollbarEndpoints.tsx` | Endcap markers at marked range boundaries |
| `useByteArray.ts` | Hook for managing a mutable `Uint8Array` with version-based change detection and optional debounce |
| `utils.ts` | Pixel-fill mapping, contiguous run detection, indicator layout math |
| `index.ts` | Barrel export |

**Modified** — `platform/ui-next/src/components/index.ts` to export the new components.

**Usage:**
```tsx
const loaded = useByteArray(imageIds.length);
const viewed = useByteArray(imageIds.length);

<SmartScrollbar value={index} total={295} onValueChange={setIndex} isLoading={loading}>
  <SmartScrollbarTrack>
    <SmartScrollbarFill marked={loaded.bytes} version={loaded.version} className="bg-neutral/25" loadingClassName="bg-neutral/50" />
    <SmartScrollbarFill marked={viewed.bytes} version={viewed.version} className="bg-primary/35" />
  </SmartScrollbarTrack>
  <SmartScrollbarIndicator />
  <SmartScrollbarEndpoints marked={loaded.bytes} version={loaded.version} />
</SmartScrollbar>
```

**Key behaviors:**
- Track contracts from 8px to 4px after loading completes (600ms settle delay, 300ms ease transition)
- Expands on hover or drag
- Dot-grid pattern fades in during loading, fades out on completion
- Loaded fill brightens during loading (50% opacity), dims after (25%)
- Endpoint caps stay stationary during contraction (rendered via portal to a stable layer)
- Click-to-jump and drag-to-scrub with pointer capture
- Fills and endpoints render in pixel space — integer coordinates prevent sub-pixel bleeding and false gaps
- Memoized subcomponents with version-based change detection for efficient re-renders
- Error thrown if `SmartScrollbarIndicator` is omitted
- Uses existing OHIF design tokens (`--neutral`, `--primary`, `--foreground`)
- Fill colors are consumer-configurable via Tailwind class props

### Testing

No Testing Needed.

This PR adds a UI component only — no viewer integration is included. The component can be previewed and tested interactively here:

**Live demo:** https://ohif-scroll-design-final.netlify.app/

- Press Play to see the loading lifecycle (dot-grid, fills, contraction)
- Scroll the viewport to move the indicator
- Hover the contracted scrollbar to see it expand
- Drag the indicator to scrub through slices

#### PR

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

- [ ] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] OS: macOS Sequoia 15.7.4 (24G517), Apple M4 Pro
- [x] Node version: 20.19.1
- [x] Browser: Google Chrome (latest) 146.0.7680.165 (Official Build) (arm64)

<h3>Greptile Summary</h3>

This PR adds a well-designed `SmartScrollbar` compound component to `@ohif/ui-next` that visualizes slice loading progress, viewed state, and scroll position. The implementation is pure UI with no Cornerstone dependencies, uses context providers for layout and scroll state, handles pointer interaction with capture, and renders fills/endpoints in pixel space to prevent sub-pixel bleeding.

Previous review concerns (stale fills with in-place mutation, ref timing for portal layer, keyboard accessibility conflicts, unused `svgIdPrefix`, fill bar overflow) have all been addressed. Two minor issues remain:

- **Missing `useMemo` in `SmartScrollbarEndpoints`**: `computePixelFilledFromMarked` and the first/last pixel scan run synchronously on every context change (including hover/drag which update `effectiveWidth`). Since `SmartScrollbarLayoutContext` is recreated whenever `effectiveWidth` changes, the endpoint component re-renders on every hover event and repeats an O(n) pass over the pixel array. For large volumes (1000+ slices), this is an unnecessary allocation and scan on every pointer-enter/leave. Wrapping with `useMemo([marked, version, trackHeight, fillPadding])` would avoid this.
- **Shallow child validation**: `validateChildren` only scans top-level children via `Children.forEach`. If a consumer wraps `SmartScrollbarIndicator` in a `React.Fragment` or conditional render (e.g., `{condition && <SmartScrollbarIndicator />}`), the check fails to detect it and throws incorrectly. This is a potential footgun for consumers.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all previous P1 concerns have been resolved; remaining findings are P2 style/performance suggestions.

All blocking issues from previous review rounds (ref timing, stale fills, keyboard conflicts, unused context fields, run length overflow) have been addressed. The two remaining findings are P2: a missing useMemo in SmartScrollbarEndpoints (performance concern for large volumes) and a shallow child scan that can throw for wrapped indicators. Neither causes incorrect behavior in the documented usage pattern.

SmartScrollbarEndpoints.tsx (missing memoization for pixel computation) and SmartScrollbar.tsx (shallow validateChildren scan).

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx | Root component with dual context providers, ResizeObserver, opt-in keyboard nav, pointer handling, and contraction state — all previous review concerns addressed. |
| platform/ui-next/src/components/SmartScrollbar/SmartScrollbarEndpoints.tsx | Portal-based endpoint cap renderer; missing useMemo for O(n) pixel computation that re-runs on every context change (hover/drag). |
| platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx | Pixel-aligned fill bars with correct useMemo and version-based invalidation; clean implementation. |
| platform/ui-next/src/components/SmartScrollbar/SmartScrollbarIndicator.tsx | Pill-shaped position indicator aligned to pixel buckets; straightforward and correct. |
| platform/ui-next/src/components/SmartScrollbar/SmartScrollbarTrack.tsx | Dot-grid background with fade animation; correctly uses useId for pattern IDs and memo for re-render control. |
| platform/ui-next/src/components/SmartScrollbar/utils.ts | Pure utility functions for pixel mapping, run detection, and indicator layout; well-tested with good edge-case coverage. |
| platform/ui-next/src/components/SmartScrollbar/utils.test.ts | Solid unit test coverage for computePixelFilledFromMarked and computeContiguousRuns including boundary conditions. |
| platform/ui-next/src/components/SmartScrollbar/useByteArray.ts | Mutable Uint8Array with version-based React invalidation, optional debounce, and well-documented API contract. |
| platform/ui-next/src/components/SmartScrollbar/index.ts | Barrel export for all SmartScrollbar components and hooks. |
| platform/ui-next/src/components/index.ts | Adds SmartScrollbar exports to the ui-next component index. |

</details>

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: platform/ui-next/src/components/SmartScrollbar/SmartScrollbarEndpoints.tsx
Line: 35-54

Comment:
**Missing `useMemo` for O(n) pixel computation**

`computePixelFilledFromMarked` and the first/last pixel scan run unconditionally on every render. `SmartScrollbarEndpoints` consumes `SmartScrollbarLayoutContext`, which is recreated whenever `effectiveWidth` changes — meaning this component re-renders (and repeats the full O(n) scan) on every hover and drag event, even though `trackHeight` and `fillPadding` (the actual inputs to the computation) haven't changed.

For a typical OHIF volume with 1000–3000 slices, that's an unnecessary allocation and scan on every pointer-enter/leave. `SmartScrollbarFill` already wraps its equivalent computation in `useMemo` — the same pattern should be applied here:

```tsx
const { firstFilledPixel, lastFilledPixel } = useMemo(() => {
  const pixelCount = Math.max(0, Math.floor(trackHeight - fillPadding * 2));
  if (pixelCount === 0) return { firstFilledPixel: -1, lastFilledPixel: -1 };
  const pixelFilled = computePixelFilledFromMarked(marked, pixelCount);
  let first = -1;
  let last = -1;
  for (let pixel = 0; pixel < pixelFilled.length; pixel++) {
    if (pixelFilled[pixel]) { first = pixel; break; }
  }
  for (let pixel = pixelFilled.length - 1; pixel >= 0; pixel--) {
    if (pixelFilled[pixel]) { last = pixel; break; }
  }
  return { firstFilledPixel: first, lastFilledPixel: last };
}, [marked, version, trackHeight, fillPadding]);
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx
Line: 16-30

Comment:
**Shallow child scan misses wrapped indicators**

`validateChildren` uses `Children.forEach` which only iterates the *direct* top-level children. If a consumer wraps `SmartScrollbarIndicator` in a `React.Fragment` or a conditional expression, the check won't find it and will throw, even though the indicator is present:

```tsx
// Both of these incorrectly throw:
<SmartScrollbar ...>
  <>
    <SmartScrollbarTrack />
    <SmartScrollbarIndicator />
  </>
</SmartScrollbar>

<SmartScrollbar ...>
  <SmartScrollbarTrack />
  {showIndicator && <SmartScrollbarIndicator />}
</SmartScrollbar>
```

The second case is especially problematic — during an initial render where `showIndicator` is `false` (e.g., waiting for data), the component crashes the React tree even though the consumer intends to add the indicator once data arrives.

Consider documenting that the indicator must be an unconditional, unwrapped direct child, or use a recursive `Children.forEach` / context-flag approach to detect it at any nesting depth.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (5): Last reviewed commit: ["Merge remote-tracking branch &#39;origin/mas..."](https://github.com/ohif/viewers/commit/c6691e7e17a69e9f058f2c358fc8e8f9dda6b6d5) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=26489295)</sub>

### CODE REVIEW

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+export interface SmartScrollbarContextValue {
+  value: number;
+  totalSlices: number;
+  trackHeight: number;
+  isLoading: boolean;
+  effectiveWidth: number;
+  trackWidth: number;
+  fillPadding: number;
+  svgIdPrefix: string;
+  stableLayerRef: RefObject<HTMLDivElement | null>;
```
[the remark this answers is not in the document]
Comment (dan-rukas) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: @greptileai Fixed. Removed `svgIdPrefix` from the context interface and component (unused dead code). The Track generates its own pattern ID locally via useId().

File: platform/ui-next/src/components/SmartScrollbar/utils.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  runs.push({ start: runStart, length: runLength, isFirst: false, isLast: false });
+
+  // Mark first and last
+  if (runs.length > 0) {
+    runs[0].isFirst = true;
+    runs[runs.length - 1].isLast = true;
+  }
+
+  // Filter to valid range
+  return runs.filter(r => r.start >= 0 && r.start < totalSlices);
```
[the remark this answers is not in the document]
Comment (dan-rukas) [reply platform/ui-next/src/components/SmartScrollbar/utils.ts]: @greptileai Fixed. Added length clamping to `getContiguousRuns`. Runs that extend past totalSlices are now truncated. Prevents fill bars from overflowing the track in edge cases.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbarEndpoints.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+interface SmartScrollbarEndpointsProps {
+  slices: Set<number>;
+  className?: string;
+}
+
+export function SmartScrollbarEndpoints({ slices, className }: SmartScrollbarEndpointsProps) {
+  const { totalSlices, trackHeight, trackWidth, fillPadding, stableLayerRef } = useSmartScrollbarContext();
+
+  if (slices.size === 0 || trackHeight === 0 || !stableLayerRef.current) return null;
```
[the remark this answers is not in the document]
Comment (dan-rukas) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbarEndpoints.tsx]: @greptile-apps Replaced `useRef` with `useState` callback ref for the stable layer element. React triggers a re-render when the DOM node mounts, so endpoints now render on the first valid pass instead of requiring a subsequent re-render. 

Fixes the case where endpoints could be invisible if `slices` never changes after the initial height measurement.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+          marginLeft: -hitZoneLeftExtension,
+          cursor: isDragging ? 'grabbing' : 'grab',
+          touchAction: 'none',
+        }}
+        onPointerEnter={() => setIsHovered(true)}
+        onPointerLeave={() => setIsHovered(false)}
+        onPointerDown={handlePointerDown}
+        onPointerMove={handlePointerMove}
+        onPointerUp={handlePointerUp}
+        onPointerCancel={handlePointerUp}
```
[the remark this answers is not in the document]
Comment (dan-rukas) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: @greptileai Added `onKeyDown` handler for full keyboard accessibility. Supports Arrow keys (step by 1), Page Up/Down (step by 10), and Home/End (jump to min/max). Ensures keyboard and assistive technology is usable with the component in the Viewer.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+export function SmartScrollbarFill({ slices, className, loadingClassName }: SmartScrollbarFillProps) {
+  const { totalSlices, trackHeight, effectiveWidth, fillPadding, isLoading } = useSmartScrollbarContext();
+
+  // slices is a mutated Set (same reference), so depend on .size to bust memo
+  const slicesSize = slices.size;
+  const runs = useMemo(
+    () => getContiguousRuns(slices, totalSlices),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [slicesSize, totalSlices]
```
[the remark this answers is not in the document]
Comment (dan-rukas) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx]: The `slices.size` dependency works for the expected data flow — loaded slices grow as the cache fills and viewed slices accumulate as the user scrolls. The edge case (remove + add in the same render cycle with size unchanged) would require React to batch a cache eviction and a cache addition into a single render. This is theoretically possible with React 18 automatic batching but very unlikely given these are separate async events.

Flagging this for @jbocce to review/evaluate further based on the actual Cornerstone event model.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+  let hasIndicator = false;
+
+  Children.forEach(children, (child) => {
+    if (!isValidElement(child)) return;
+    if (child.type === SmartScrollbarIndicator) hasIndicator = true;
+  });
+
+  if (!hasIndicator && !_warnedNoIndicator) {
+    _warnedNoIndicator = true;
+    console.warn(
```
Comment (jbocce) [inline platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: I strongly believe that we should throw an exception here and no need to check if we already warned and/or threw an exception already.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbocce) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: @greptileai and @dan-rukas  we need to proceed with caution here. I am wondering if these could/will conflict with the various keyboard shortcuts in OHIF already? If this component were to be something to be used outside of OHIF then I would agree with these changes, but I am not so sure about adding these in. Thoughts? At the every list we could make it an option that is off by default and that could be turned on. So for example add a boolean property to the `SmartScrollbar` component (say `supportWaiAria`) or something?

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbocce) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx]: Thanks @greptileai, you make a good point here. This becomes even more critical because we will be switching from using a `Set<number>` to a `UInt8Array` to keep track of the slices. It is also a point well taken because it is possible for slices to be unloaded/uncached. As part of that change to `UInt8Array` we will keep this point in mind and do something like the following...

```
const prevSnapshotRef = useRef<Uint8Array | null>(null);
const versionRef = useRef(0);
// Runs every render — bumps version only when content actually changed
if (
  prevSnapshotRef.current === null ||
  prevSnapshotRef.current.length !== slices.length ||
  prevSnapshotRef.current.some((v, i) => v !== slices[i])
) {
  prevSnapshotRef.current = slices.slice();
  versionRef.current++;
}
const runs = useMemo(
  () => getContiguousRuns(slices, totalSlices),
  [versionRef.current, totalSlices]
);
```

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbocce) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: @greptileai, I just pushed a resolution to this. Look good?

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+      const ratio = Math.max(0, Math.min(1, (clientY - trackTopRef.current) / trackHeight));
+      return Math.round(ratio * (totalSlices - 1));
+    },
+    [trackHeight, totalSlices]
+  );
+
+  const handlePointerDown = useCallback(
+    (e: React.PointerEvent) => {
+      const trackEl =
+        (e.currentTarget as HTMLElement).querySelector('[data-scrollbar-track]') as HTMLElement
```
Comment (jbocce) [inline platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: @greptileai this use of the `data-scrollbar-track` attribute concerns me a bit. It is not very React idiomatic. Wouldn't a useRef be better suited here?

Comment (jbocce) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: @greptileai, I don't think the inner track needs to be referenced at all. I just pushed a commit to address this. Have a look.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+      }
+
+      e.preventDefault();
+      onValueChange(clamp(next));
+    },
+    [value, totalSlices, clamp, onValueChange]
+  );
+
+  // ── Context value ────────────────────────────────────────────
+  const ctx: SmartScrollbarContextValue = {
```
Comment (jbocce) [inline platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: Here we have a combined context which makes things easier to read and understand from a code standpoint, but I think because in particular the `value` property is mixed in here, all the more "static" scroll bar sub-components will get re-rendered every time value changes (i.e. scrolling occurs) when they don't need to be. I think it is best to split the context into two - value and layout. I will look at this next.

FYI @greptileai

Comment (jbocce) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: I just pushed a commit @greptileai to address this. How does it look?

**Comment by jbocce** [comment]
The `SmartScrollbar` component is supposed to be somewhat generic. Eliminating references to slices would be prudent. I will push a commit to address this soon.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx (hunk 6 above)
Comment (dan-rukas) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx]: Agreed! Updated - switched to a throw and removed the warning.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbocce) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx]: @greptileai, the change I was suggesting was pushed yesterday I think. Any thoughts?

**Comment by jbocce** [comment]
> The `SmartScrollbar` component is supposed to be somewhat generic. Eliminating references to slices would be prudent. I will push a commit to address this soon.

Commit hash a3880846114e5839f8ff1930fc5a6f40b5090921 addressed this.

**Comment by jbocce** [comment]
@greptileai I just pushed bc80fae7be65cbdb5a1effcbd6ade5b0fee5b87b that addresses a concern we had for ensuring the fills accurately indicate marked items without over and/or under stating items that are marked. This is especially important considering that one fill is used to indicate viewed (and unviewed) medical images. Please have a look and let me know what you think. Thanks.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        'Users will not see their current scroll position without it.'
+    );
+  }
+}
+
+// ── Layout and timing constants ─────────────────────────────────
+const TRACK_WIDTH = 8;
+const RESTING_WIDTH = 4;
+const FILL_PADDING = 3;
+const INDICATOR_SIZE = 8;
```
[the remark this answers is not in the document]
Comment (jbocce) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbar.tsx:36]: @greptileai I changed the PR description to match the code.

File: platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbocce) [reply platform/ui-next/src/components/SmartScrollbar/SmartScrollbarFill.tsx]: @greptileai I pushed a commit with JSDocs to clarify things for developers/consumers.

**Comment by jbocce** [comment]
@greptileai I just pushed some unit tests for some of the SmartSrollbar utilities. Please have a look and let me know if there are any you think I should add. Thanks.

**Review by jbocce — APPROVED** [review]

**Comment by dan-rukas** [comment]
Tested the latest code and changes and everything looks great and matches the original design and behaviors. Thanks again @jbocce for your help, changes, and detailed approach to getting this right.
